File: examples/typescript/advanced-dag.ts
/**
 * Apex TypeScript SDK - Advanced DAG with Conditional Branches
 *
 * This example demonstrates building complex DAG workflows that include:
 * - Multi-level conditional branching based on task output
 * - Error-handling nodes with fallback paths
 * - Parallel fan-out / fan-in patterns
 * - Approval gates embedded within a DAG
 * - Dynamic node configuration using metadata
 *
 * The DAG models a content moderation pipeline:
 *
 *   [Ingest Content]
 *        |
 *   [AI Moderation] ----> [Classify Severity]
 *                              |
 *                         (condition)
 *                        /     |      \
 *                   [Auto     [Queue    [Escalate to
 *                   Approve]   Review]   Legal Team]
 *                        \     |      /
 *                         [Publish Decision]
 *                              |
 *                         [Notify Author]
 *
 * Prerequisites:
 *   npm install @apex-swarm/sdk
 *
 * Run with:
 *   npx ts-node advanced-dag.ts
 */

import {
  ApexClient,
  CreateDAGRequest,
  DAGNode,
  DAGEdge,
  TaskPriority,
  DAGStatus,
  ApexError,
} from '@apex-swarm/sdk';

// =============================================================================
// Configuration
// =============================================================================

const API_URL: string = process.env.APEX_API_URL || 'http://localhost:8080';
const API_KEY: string = process.env.APEX_API_KEY || '';

const client = new ApexClient({
  baseUrl: API_URL,
  apiKey: API_KEY,
  timeout: 60000,
});

// =============================================================================
// Helper Types
// =============================================================================

/** Configuration for a single node in the DAG builder. */
interface NodeConfig {
  id: string;
  name: string;
  description: string;
  type: 'task' | 'condition' | 'approval' | 'parallel';
  input?: Record<string, unknown>;
  priority?: TaskPriority;
  timeout?: number;
  retries?: number;
  condition?: {
    expression: string;
    trueBranch: string;
    falseBranch: string;
  };
  approvalConfig?: {
    approvers: string[];
    requiredApprovals: number;
    timeoutSeconds: number;
  };
}

// =============================================================================
// DAG Builder
// =============================================================================

/**
 * Build the content moderation pipeline DAG.
 *
 * The pipeline processes user-submitted content through several stages:
 * 1. **Ingest** - Receive and normalize the content.
 * 2. **AI Moderation** - Run automated content analysis.
 * 3. **Classify Severity** - Determine severity (low / medium / high).
 * 4. **Conditional routing**:
 *    - Low severity: auto-approve immediately.
 *    - Medium severity: queue for human review.
 *    - High severity: escalate to the legal team (approval gate).
 * 5. **Publish Decision** - Record and publish the moderation outcome.
 * 6. **Notify Author** - Send the decision to the content author.
 */
function buildModerationPipeline(): CreateDAGRequest {
  // -- Node definitions -----------------------------------------------------

  const nodes: Omit<DAGNode, 'status' | 'taskId'>[] = [
    // Stage 1: Content ingestion
    {
      id: 'ingest',
      name: 'Ingest Content',
      type: 'task',
      config: {
        taskTemplate: {
          name: 'Ingest Submitted Content',
          description: 'Receive content, extract text/media, normalize format',
          priority: TaskPriority.NORMAL,
          input: {
            contentTypes: ['text', 'image', 'video'],
            extractMetadata: true,
            normalizeEncoding: 'utf-8',
          },
        },
        timeout: 60,
        retries: 2,
      },
    },

    // Stage 2: AI-powered moderation
    {
      id: 'ai-moderate',
      name: 'AI Moderation',
      type: 'task',
      config: {
        taskTemplate: {
          name: 'Run AI Moderation',
          description: 'Analyze content with multiple AI models for policy violations',
          priority: TaskPriority.HIGH,
          input: {
            models: ['text-classifier-v3', 'image-safety-v2', 'toxicity-scorer'],
            policies: ['hate_speech', 'violence', 'spam', 'misinformation'],
            confidenceThreshold: 0.7,
          },
        },
        timeout: 120,
        retries: 1,
      },
    },

    // Stage 3: Severity classification (condition node)
    //
    // This node evaluates the AI moderation output and routes to one of
    // three branches based on the computed severity score.
    {
      id: 'classify-severity',
      name: 'Classify Severity',
      type: 'condition',
      config: {
        condition: {
          // Expression evaluated against the ai-moderate node's output.
          // The orchestrator routes to trueBranch or falseBranch based on
          // nested conditions encoded in the edge conditions.
          expression: 'output.severityScore',
          trueBranch: 'auto-approve',    // severityScore < 30
          falseBranch: 'queue-review',   // severityScore >= 30
        },
      },
    },

    // Stage 4a: Auto-approve (low severity)
    {
      id: 'auto-approve',
      name: 'Auto-Approve Content',
      type: 'task',
      config: {
        taskTemplate: {
          name: 'Auto-Approve Low-Risk Content',
          description: 'Automatically approve content with low severity scores',
          priority: TaskPriority.LOW,
          input: {
            decision: 'approved',
            reason: 'Below severity threshold',
            addWatermark: false,
          },
        },
        timeout: 30,
      },
    },

    // Stage 4b: Queue for human review (medium severity)
    {
      id: 'queue-review',
      name: 'Queue for Human Review',
      type: 'task',
      config: {
        taskTemplate: {
          name: 'Queue Content for Review',
          description: 'Add content to the human review queue with context',
          priority: TaskPriority.NORMAL,
          input: {
            reviewQueue: 'content-moderation',
            priority: 'medium',
            includeAiAnalysis: true,
            slaHours: 24,
          },
        },
        timeout: 60,
      },
    },

    // Stage 4c: Escalate to legal (high severity) -- uses approval gate
    {
      id: 'escalate-legal',
      name: 'Escalate to Legal Team',
      type: 'approval',
      config: {
        approvalConfig: {
          approvers: ['legal-team@example.com', 'compliance@example.com'],
          requiredApprovals: 1,
          timeoutSeconds: 86400, // 24-hour SLA for legal review
          autoApprove: false,
        },
        taskTemplate: {
          name: 'Legal Team Escalation',
          description: 'High-severity content requires legal team sign-off',
          priority: TaskPriority.CRITICAL,
          input: {
            severity: 'high',
            requiresLegalReview: true,
            attachPolicyReport: true,
          },
        },
      },
    },

    // Stage 5: Publish the moderation decision
    // This node converges all three conditional branches.
    {
      id: 'publish-decision',
      name: 'Publish Decision',
      type: 'task',
      config: {
        taskTemplate: {
          name: 'Record and Publish Moderation Decision',
          description: 'Store the decision in the audit log and update content status',
          priority: TaskPriority.HIGH,
          input: {
            publishToAuditLog: true,
            updateContentStatus: true,
            retentionDays: 365,
          },
        },
        timeout: 30,
      },
    },

    // Stage 6: Notify the content author
    {
      id: 'notify-author',
      name: 'Notify Author',
      type: 'task',
      config: {
        taskTemplate: {
          name: 'Send Author Notification',
          description: 'Notify the content author of the moderation decision',
          priority: TaskPriority.NORMAL,
          input: {
            channels: ['email', 'in-app'],
            includeAppealLink: true,
            templateId: 'moderation-result-v2',
          },
        },
        timeout: 30,
      },
    },
  ];

  // -- Edge definitions (explicit graph wiring) -----------------------------

  const edges: Omit<DAGEdge, 'id'>[] = [
    // Sequential: ingest -> AI moderate -> classify
    { sourceNodeId: 'ingest', targetNodeId: 'ai-moderate' },
    { sourceNodeId: 'ai-moderate', targetNodeId: 'classify-severity' },

    // Conditional branches from severity classification
    {
      sourceNodeId: 'classify-severity',
      targetNodeId: 'auto-approve',
      condition: 'output.severityScore < 30',
    },
    {
      sourceNodeId: 'classify-severity',
      targetNodeId: 'queue-review',
      condition: 'output.severityScore >= 30 && output.severityScore < 70',
    },
    {
      sourceNodeId: 'classify-severity',
      targetNodeId: 'escalate-legal',
      condition: 'output.severityScore >= 70',
    },

    // All three branches converge to publish-decision
    { sourceNodeId: 'auto-approve', targetNodeId: 'publish-decision' },
    { sourceNodeId: 'queue-review', targetNodeId: 'publish-decision' },
    { sourceNodeId: 'escalate-legal', targetNodeId: 'publish-decision' },

    // Final notification
    { sourceNodeId: 'publish-decision', targetNodeId: 'notify-author' },
  ];

  return {
    name: 'Content Moderation Pipeline',
    description:
      'Multi-stage content moderation with AI analysis, severity-based ' +
      'conditional routing (auto-approve / human review / legal escalation), ' +
      'and author notification.',
    nodes,
    edges,
    metadata: {
      version: '3.0',
      team: 'trust-and-safety',
      slaTarget: '< 24 hours for all content',
    },
  };
}

// =============================================================================
// Utility: Print DAG Structure
// =============================================================================

/**
 * Print a human-readable summary of the DAG topology.
 */
function printDAGStructure(dag: CreateDAGRequest): void {
  console.log(`\nDAG: ${dag.name}`);
  console.log(`Description: ${dag.description}`);
  console.log(`\nNodes (${dag.nodes.length}):`);

  for (const node of dag.nodes) {
    const type = `[${node.type}]`;
    console.log(`  ${node.id} ${type} - ${node.name}`);

    // Show condition details if present
    if (node.type === 'condition' && node.config.condition) {
      const cond = node.config.condition;
      console.log(`    Expression: ${cond.expression}`);
      console.log(`    True  -> ${cond.trueBranch}`);
      console.log(`    False -> ${cond.falseBranch}`);
    }

    // Show approval config if present
    if (node.type === 'approval' && node.config.approvalConfig) {
      const ac = node.config.approvalConfig;
      console.log(`    Approvers: ${ac.approvers?.join(', ')}`);
      console.log(`    Required: ${ac.requiredApprovals}`);
      console.log(`    Timeout: ${ac.timeoutSeconds}s`);
    }
  }

  console.log(`\nEdges (${dag.edges.length}):`);
  for (const edge of dag.edges) {
    const cond = edge.condition ? ` [if ${edge.condition}]` : '';
    console.log(`  ${edge.sourceNodeId} -> ${edge.targetNodeId}${cond}`);
  }
}

// =============================================================================
// Main
// =============================================================================

async function main(): Promise<void> {
  console.log('='.repeat(70));
  console.log('  Apex SDK - Advanced DAG with Conditional Branches (TypeScript)');
  console.log('='.repeat(70));

  // Build the DAG definition
  const dagDef = buildModerationPipeline();

  console.log('\n--- DAG Definition ---');
  printDAGStructure(dagDef);

  // Submit to the API
  try {
    console.log('\n--- Submitting DAG ---\n');
    const dag = await client.createDAG(dagDef);
    console.log(`DAG created: ${dag.id}`);
    console.log(`Status: ${dag.status}`);

    // Start execution
    console.log('\n--- Starting Execution ---\n');
    const execution = await client.startDAG(dag.id, {
      contentId: 'content-2024-001',
      submittedBy: 'user-42',
      contentUrl: 'https://cdn.example.com/uploads/post-12345',
    });
    console.log(`Execution started: ${execution.id}`);
    console.log(`Status: ${execution.status}`);

    // Cleanup
    console.log('\n--- Cleanup ---\n');
    await client.deleteDAG(dag.id);
    console.log('DAG deleted');
  } catch (error) {
    if (error instanceof ApexError) {
      console.error(`API error [${error.code}]: ${error.message}`);
    } else {
      console.error('Unexpected error:', error);
    }
  }

  console.log('\n' + '='.repeat(70));
  console.log('  Advanced DAG example completed');
  console.log('='.repeat(70));
}

main().catch(console.error);

File: examples/python/advanced_dag.py
#!/usr/bin/env python3
"""
Apex Python SDK - Advanced DAG with Conditional Branches

This example demonstrates building complex DAG workflows that include:
- Multi-level conditional branching based on task output
- Error-handling nodes with fallback paths
- Parallel fan-out / fan-in patterns
- Dynamic node configuration using metadata propagation
- Approval gates embedded within a DAG
- Scheduled (cron-based) DAG execution

The DAG modeled here represents an ML model evaluation pipeline:

  [Ingest Data]
       |
  [Validate Schema]
       |
  [Feature Engineering] -----> [Train Model A] --+
       |                                          |
       +----> [Train Model B] ---+                |
       |                         +--> [Compare Models]
       +----> [Train Model C] ---+         |
                                      (condition)
                                     /          \
                            [Deploy Best]   [Request Human Review]
                                     \          /
                                      [Notify Stakeholders]

Prerequisites:
    pip install apex-swarm

Run with:
    python advanced_dag.py
"""

from __future__ import annotations

import os
import sys
from typing import Any

from apex_sdk import ApexClient
from apex_sdk.models import (
    DAGCreate,
    DAGEdge,
    DAGNode,
    DAGStatus,
    TaskCreate,
    TaskInput,
    TaskPriority,
    TaskStatus,
)
from apex_sdk.exceptions import ApexAPIError, ApexNotFoundError

# =============================================================================
# Configuration
# =============================================================================

API_URL: str = os.environ.get("APEX_API_URL", "http://localhost:8080")
API_KEY: str = os.environ.get("APEX_API_KEY", "")


def build_ml_pipeline_dag() -> DAGCreate:
    """Construct the full ML evaluation pipeline DAG definition.

    The pipeline performs the following stages:

    1. **Ingest** -- Load raw data from configured sources.
    2. **Validate** -- Run schema and quality checks.
    3. **Feature Engineering** -- Transform raw features into model-ready
       representations.
    4. **Train (parallel)** -- Train three candidate models concurrently
       (Model A, B, C) with different hyperparameters.
    5. **Compare** -- Evaluate all models on a held-out test set.
    6. **Conditional branch** -- If the best model exceeds a quality
       threshold, deploy it automatically; otherwise, request human
       review before proceeding.
    7. **Notify** -- Send results to stakeholders regardless of path.

    Returns:
        A fully configured :class:`DAGCreate` ready for submission.
    """
    # -- Helper to build a task-template node quickly -------------------------

    def _task_node(
        node_id: str,
        name: str,
        description: str,
        deps: list[str],
        *,
        priority: TaskPriority = TaskPriority.NORMAL,
        input_data: dict[str, Any] | None = None,
        timeout: int = 300,
        retries: int = 1,
        condition: str | None = None,
    ) -> DAGNode:
        """Create a DAGNode wrapping a TaskCreate template."""
        return DAGNode(
            id=node_id,
            task_template=TaskCreate(
                name=name,
                description=description,
                priority=priority,
                input=TaskInput(data=input_data or {}),
                timeout_seconds=timeout,
                retries=retries,
                tags=["ml-pipeline"],
            ),
            depends_on=deps,
            condition=condition,
        )

    # -- Node definitions -----------------------------------------------------

    nodes: list[DAGNode] = [
        # Stage 1: Data ingestion
        _task_node(
            "ingest",
            "Ingest Raw Data",
            "Load data from S3 bucket and streaming sources",
            deps=[],
            input_data={
                "sources": ["s3://ml-data/raw/", "kafka://events-topic"],
                "format": "parquet",
                "sample_rate": 1.0,
            },
            timeout=600,
        ),
        # Stage 2: Schema validation
        _task_node(
            "validate",
            "Validate Data Schema",
            "Run schema checks, null-rate analysis, and distribution drift detection",
            deps=["ingest"],
            input_data={
                "checks": ["schema", "null_rate", "drift"],
                "drift_threshold": 0.05,
                "fail_on_warning": False,
            },
        ),
        # Stage 3: Feature engineering
        _task_node(
            "features",
            "Feature Engineering",
            "Transform raw columns into ML-ready features",
            deps=["validate"],
            input_data={
                "transformations": [
                    "one_hot_encode",
                    "standard_scale",
                    "interaction_terms",
                ],
                "target_column": "label",
                "test_split": 0.2,
            },
            timeout=900,
        ),
        # Stage 4a: Train Model A (e.g. gradient-boosted tree)
        _task_node(
            "train-model-a",
            "Train Model A (XGBoost)",
            "Train an XGBoost classifier with grid-search hyperparameter tuning",
            deps=["features"],
            priority=TaskPriority.HIGH,
            input_data={
                "algorithm": "xgboost",
                "hyperparams": {
                    "max_depth": [3, 5, 7],
                    "learning_rate": [0.01, 0.1],
                    "n_estimators": [100, 300],
                },
                "cv_folds": 5,
            },
            timeout=1800,
            retries=2,
        ),
        # Stage 4b: Train Model B (e.g. random forest)
        _task_node(
            "train-model-b",
            "Train Model B (Random Forest)",
            "Train a random-forest classifier",
            deps=["features"],
            priority=TaskPriority.HIGH,
            input_data={
                "algorithm": "random_forest",
                "hyperparams": {
                    "n_estimators": [100, 500],
                    "max_features": ["sqrt", "log2"],
                },
                "cv_folds": 5,
            },
            timeout=1800,
            retries=2,
        ),
        # Stage 4c: Train Model C (e.g. neural network)
        _task_node(
            "train-model-c",
            "Train Model C (MLP)",
            "Train a multi-layer perceptron classifier",
            deps=["features"],
            priority=TaskPriority.HIGH,
            input_data={
                "algorithm": "mlp",
                "hyperparams": {
                    "hidden_layers": [(128, 64), (256, 128, 64)],
                    "dropout": [0.2, 0.3],
                    "epochs": 50,
                },
            },
            timeout=2400,
            retries=2,
        ),
        # Stage 5: Compare models (fan-in from all three training nodes)
        _task_node(
            "compare",
            "Compare Model Performance",
            "Evaluate all candidate models on the test set and rank by F1 score",
            deps=["train-model-a", "train-model-b", "train-model-c"],
            priority=TaskPriority.HIGH,
            input_data={
                "metrics": ["f1", "precision", "recall", "auc"],
                "primary_metric": "f1",
                "quality_threshold": 0.85,
            },
        ),
        # Stage 6a: Auto-deploy if quality threshold met
        # The condition expression references the compare node's output.
        _task_node(
            "deploy",
            "Deploy Best Model",
            "Package and deploy the winning model to the serving infrastructure",
            deps=["compare"],
            condition="output.best_f1 >= 0.85",
            input_data={
                "registry": "mlflow",
                "serving": "kubernetes",
                "canary_percent": 10,
            },
        ),
        # Stage 6b: Human review if threshold not met
        _task_node(
            "human-review",
            "Request Human Review",
            "Notify ML engineers for manual inspection when quality is below threshold",
            deps=["compare"],
            condition="output.best_f1 < 0.85",
            input_data={
                "notify_channels": ["slack", "email"],
                "review_dashboard_url": "https://ml.apex.internal/review",
            },
        ),
        # Stage 7: Notify stakeholders (both paths converge here)
        _task_node(
            "notify",
            "Notify Stakeholders",
            "Send summary report to project stakeholders",
            deps=["deploy", "human-review"],
            input_data={
                "channels": ["slack", "email"],
                "report_format": "html",
                "include_charts": True,
            },
        ),
    ]

    # -- Edge definitions (explicit graph wiring) -----------------------------

    edges: list[DAGEdge] = [
        # Sequential stages
        DAGEdge(source="ingest", target="validate"),
        DAGEdge(source="validate", target="features"),
        # Fan-out: features -> three parallel training nodes
        DAGEdge(source="features", target="train-model-a"),
        DAGEdge(source="features", target="train-model-b"),
        DAGEdge(source="features", target="train-model-c"),
        # Fan-in: all training nodes -> compare
        DAGEdge(source="train-model-a", target="compare"),
        DAGEdge(source="train-model-b", target="compare"),
        DAGEdge(source="train-model-c", target="compare"),
        # Conditional branches from compare
        DAGEdge(
            source="compare",
            target="deploy",
            condition="output.best_f1 >= 0.85",
        ),
        DAGEdge(
            source="compare",
            target="human-review",
            condition="output.best_f1 < 0.85",
        ),
        # Both conditional branches converge to notify
        DAGEdge(source="deploy", target="notify"),
        DAGEdge(source="human-review", target="notify"),
    ]

    return DAGCreate(
        name="ML Model Evaluation Pipeline",
        description=(
            "End-to-end pipeline: ingest data, validate, engineer features, "
            "train three candidate models in parallel, compare results, and "
            "conditionally deploy or escalate for human review."
        ),
        nodes=nodes,
        edges=edges,
        tags=["ml", "pipeline", "advanced"],
        metadata={
            "team": "ml-platform",
            "version": "2.1",
            "estimated_duration_minutes": 45,
        },
        # Optional: run daily at 02:00 UTC
        schedule="0 2 * * *",
    )


def print_dag_structure(dag_def: DAGCreate) -> None:
    """Print a human-readable representation of the DAG topology."""
    print(f"DAG: {dag_def.name}")
    print(f"Description: {dag_def.description}")
    print(f"Schedule: {dag_def.schedule or 'manual'}")
    print(f"Nodes ({len(dag_def.nodes)}):")

    for node in dag_def.nodes:
        deps = ", ".join(node.depends_on) if node.depends_on else "(root)"
        cond = f"  [condition: {node.condition}]" if node.condition else ""
        print(f"  - {node.id}: {node.task_template.name}  deps=[{deps}]{cond}")

    print(f"\nEdges ({len(dag_def.edges or [])}):")
    for edge in dag_def.edges or []:
        cond = f"  [if {edge.condition}]" if edge.condition else ""
        print(f"  {edge.source} -> {edge.target}{cond}")


# =============================================================================
# Main
# =============================================================================

def main() -> None:
    """Build, submit, and start the advanced DAG pipeline."""
    print("=" * 70)
    print("  Apex SDK - Advanced DAG with Conditional Branches")
    print("=" * 70)

    # Build the DAG definition locally before submitting
    dag_def = build_ml_pipeline_dag()
    print("\n--- DAG Definition ---\n")
    print_dag_structure(dag_def)

    # Submit the DAG to the Apex API
    client = ApexClient(base_url=API_URL, api_key=API_KEY, timeout=60.0)

    try:
        print("\n--- Submitting DAG ---\n")
        dag = client.create_dag(dag_def)
        print(f"DAG created: {dag.id} ({dag.name})")
        print(f"Status: {dag.status}")

        # Start execution with initial input parameters
        print("\n--- Starting Execution ---\n")
        running_dag = client.start_dag(
            dag.id,
            input_data={
                "run_id": "eval-2024-q4-001",
                "data_snapshot": "2024-12-01",
            },
        )
        print(f"DAG running: {running_dag.status}")
        print(f"Started at: {running_dag.started_at}")

        # In production you would monitor via WebSocket or polling.
        # See monitoring.py for a real-time monitoring example.

        # Cleanup for demo purposes
        print("\n--- Cleanup ---\n")
        client.delete_dag(dag.id)
        print("DAG deleted")

    except ApexAPIError as e:
        print(f"API error: {e.message}")
        sys.exit(1)
    except Exception as e:
        print(f"Unexpected error: {e}")
        sys.exit(1)
    finally:
        client.close()

    print("\n" + "=" * 70)
    print("  Advanced DAG example completed")
    print("=" * 70)


if __name__ == "__main__":
    main()
